### DISCUSSION TITLE
Native support for matrix/array operations avoiding the creation of  temporaries

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6893] [Language Ideas]
Dear all,
much shorter code, much more maintainable, much more clear and easy to follow and self-commented in several applications where mathematical problems are needed to be solved using C#, would be allowed provided that native support for expressions like the following is natively supported in C# as it is the case with Fortran, Matlab, Julia
```C#
x = alpha*y + beta*z + gamma*(w - z*y);
```
x, y, z, w are C# `double[]` or `float[]` or `int[]` arrays  while alpha, beta, gamma are of type `double` or `float` or `int` . The C# compiler is expected to replace this expression with the hand coded loop below
```C#
for (int i = 0; i < y.Length; i++)
{
   x[i] = alpha*y[i] + beta*z[i] + gamma*(w[i] - z[i]*y[i]);
}
```
 in order to avoid the creation of temporary objects, introduced in other languages where the same functionality is achieved by operator overloading. Temporaries occupy memory and have a penalty of allocation / deallocation. Fortran supports this feature by design and this is the only reason people recommend Fortran or Matlab or Julia for Numerical libraries solving large computer simulations whether it is molecular dynamics, or weather prediction, or aeronautics, and several other problems based on computational physics. Why not C# with all these beautiful features? Fortran also allows dense matrices in such expressions. So we would like to have dense matrices of type `double[,]` lets say A, B for example and expressions like
 ```C#
x = alpha*A*y + beta*z + gamma*B*(w - z*y);
```
Thank you very much for your time and effort. Do you think that it would be nice to support this functionality in a modern language like C# instead of trying to reproduce it with generics or other inheritance tricks since older languages support it from since several decades ago? Further more there should be some interface allowing the user to define his own types, quite commonly such operations are needed for Sparse Matrices (sparse matrix times vector products), and define a matrix vector product with specific keywords allowing the compiler to allow such expressions without introducing temporaries. Consider the previous expression where A, B are user defined structs or classes.

**Comment by orthoxerox** [comment]
You can do this by making a struct (to avoid allocations) wrapper around your arrays. This will allow you to overload the operators you want.

There's also #5496, which will allow something similar with more syntactic sugar if it's implemented.

**Reply by ghost** [reply]
You end up with the same problems as before. You do not avoid temporary objects.

**Reply by orthoxerox** [reply]
I do. A struct wrapper is literally a compile-time thing, not a single extra byte is allocated.

**Reply by ghost** [reply]
`x1+x2+x3 + x4 = x1+x2+temp1 = x1 + temp2`
temp1 is a temporary where `temp1=x3+x4`
temp2 is another temporary object where `temp2 = x2+temp1`
So, by overloading operators you will end up with the creation of temporary arrays holding intermediate results. If the arrays are large 10^8 entries you need 800MB for each temporary. There is not only runtime overhead but memory overhead.

**Reply by orthoxerox** [reply]
Ah, I see what you're trying to do. I think this is called loop fusion and optimizations like this belong to the runtime, not the language.

**Reply by ghost** [reply]
Fortran and other languages such as Matlab and Julia, supports this by design. In C++ you can do it by using template metaprogramming techniques. Since Fortran has it for more than 30 years ago, I was expecting a modern Language such as C# would support it by design, or since there is an intermediate layer such operations on native C# arrays could be translated at first to the hand coded loop
```C#
for (int i = 0; i < y.Length; i++)
{
   x[i] = alpha*y[i] + beta*z[i] + gamma*(w[i] - z[i]*y[i]);
}
```

**Reply by bernd5** [reply]
See: https://github.com/dotnet/csharplang/discussions/6892#discussioncomment-4724822

You could defer the actual calculation and produce expressions instead.

Then finally when the result is needed the calculation can happen with optimizations (e.g. scalar multiplication can be executed first: vector * scalar1 * scalar2: can be calcuated as: (scalar1 * scalar2) * vector).
(maybe triggered by an implicit conversion operator)

In the past I had some projects with heavy matrix & vector calculations for image processing.
I used emguCV which is a wrapper around OpenCV - AFAIK they defer the actual calculation to avoid temporary matrixes / vectors... (but not sure)

**Reply by ghost** [reply]
I have such a version that uses Expression inheriting from an interface, but then the problem is that the expression member function .Value(i) is not inlined and results in a penalty at runtime of about 10 times slower than the hand coded loop. I can't think of a way to detour this problem with inlining.

**Reply by bernd5** [reply]
Could you post the code?

**Reply by ghost** [reply]
Below you can find the main idea, without properties needed for data indexers and initialization of the arrays with arbitrary values. But the main idea is there although it needs improvements. For instance 
```C#
    public interface IArray
    {
        double Value(int i);
        public static IArray operator +(IArray L, IArray R) => new Plus<IArray, IArray>(L, R);
    }
    
    public struct Vector : IArray
    {       
        double[] data;
        double Value(int i)
        { return data[i]; } 
        public Vector(int n)
        { data = new double[n]; }
    }

    public struct Plus<Left, Right> : IArray
        where Left  : IArray 
        where Right : IArray
    {
        public readonly Left  L;
        public readonly Right R;

        public Plus(Left left, Right right)
        {
            L  = left;
            R = right;
        }
        public  double Value(int i)
        {
            return L.Value(i) + R.Value(i);
        }
    }

    class Program
    {
        static void Main()
        {
            Stopwatch stopwatch = new Stopwatch();
            int n = 100000000;
            IArray a = new Vector(n, "a");
            IArray b = new Vector(n, "b");
            IArray c = new Vector(n, "c");
            stopwatch.Reset();
            stopwatch.Start();
            IArray x = a + b + c + c;
            double sum = 0.0; 
            for (int i = 0; i < n; i++)
            {
                sum += x[i];
            }
            stopwatch.Stop();
            Console.WriteLine("Expression sum(a+b+c+c) = {0,25:e16} computed in {1,10:f2}", sum, stopwatch.ElapsedMilliseconds);  
       }
   }       
```
Of course one has to initialize the Vectors first with some values, and add the hand-coded loop to compare timings. At the moment we are not discussing whether `x` should be `Vector` or `IArray`, although it is of interest to be `Vector` with appropriate implicit operators etc, but we are interested how and why we get a performance penalty  greater than 4x due to the non-inlining of the `Value(int i)` member functions.

**Reply by bernd5** [reply]
Basically you should seperate your model and the evaluation of that model.
I assume the reason that the execution speed is reduced is that for every value access an interface indirection must be executed - which is not expensive but of course more expansive than a direct memory access. And because you do it 400.000.000 times it is quite important.

To avoid the overhead of interface method dispatching you could use static virtual (a bit more complicated but very fast):
```csharp
using System;
using System.Diagnostics;
using System.Numerics;

public interface IIndexable<Result, Index, TSelf>
    where TSelf : IIndexable<Result, Index, TSelf>

{
    public static abstract Result Value(TSelf indexable, Index index);
}

public struct Vector : IIndexable<double, int, Vector>
{
    double[] data;
    public double Value(int i)
    {
        return data[i];
    }
    public Vector(int n)
    {
        data = new double[n];
    }

    public Vector(double[] data)
    {
        this.data = data;
    }

    static double IIndexable<double, int, Vector>.Value(Vector v, int index)
    {
        return v.data[index];
    }
}

public struct Plus<Left, Right, Index, Result> : IIndexable<Result, Index, Plus<Left, Right, Index, Result>>
    where Left : IIndexable<Result, Index, Left>
    where Right : IIndexable<Result, Index, Right>
    where Result : IAdditionOperators<Result, Result, Result>
{
    public readonly Left L;
    public readonly Right R;

    public Plus(Left left, Right right)
    {
        L = left;
        R = right;
    }

    static Result IIndexable<Result, Index, Plus<Left, Right, Index, Result>>.Value(
        Plus<Left, Right, Index, Result> v, Index index)
    {
        var l = Left.Value(v.L, index);
        var r = Right.Value(v.R, index);

        return l + r;
    }
}

class Program
{
    public static Plus<Me, Other, Index, Result> AddGeneric<Result, Index, Me, Other>(Me me, Other other)
        where Me : IIndexable<Result, Index, Me>
        where Other : IIndexable<Result, Index, Other>
        where Result : IAdditionOperators<Result, Result, Result>
    {
        return new Plus<Me, Other, Index, Result>(me, other);
    }

    public static Plus<Me, Other, int, double> Add<Me, Other>(Me me, Other other)
        where Me : IIndexable<double, int, Me>
        where Other : IIndexable<double, int, Other>
    {
        return AddGeneric<double, int, Me, Other>(me, other);
    }

    public static Result GetValueGeneric<Indexable, Result, Index>(Indexable indexable, Index index)
        where Indexable : IIndexable<Result, Index, Indexable>
        where Result : IAdditionOperators<Result, Result, Result>
    {
        return Indexable.Value(indexable, index);
    }

    public static double GetValue<Indexable>(Indexable indexable, int index)
        where Indexable : IIndexable<double, int, Indexable>
    {
        return GetValueGeneric<Indexable, double, int>(indexable, index);
    }

    static void Main()
    {
        var watch = new Stopwatch();
        watch.Start();
        int n = 1;
        var a = new Vector(n);
        var b = new Vector(n);
        var c = new Vector(n);

        var x = Add(Add(Add(a, b), c), c);
        double sum = 0.0;
        for (int i = 0; i < n; i++)
        {
            sum += GetValue(x, i);
        }
        watch.Stop();
        Console.WriteLine(watch.ElapsedMilliseconds);
    }
}
```

**Reply by ghost** [reply]
Thank you very much for your time and effort. I implemented and run the above code with Visual Studio 2022, .NET 7.0, optimizations turned on, and the expression `a+b+c*c*a*b`, allowed with some modifications for `Mul` and `IMultiplyOperators`,  takes for n=10,000,000 560 msecs and for n=100,000,000 it needs 4913. The handcoded versions need 60msecs and 640 msecs correspondingly. But similar running times I have with the approach I described previously so I do not understand what are we saving using this syntax. Is it possible that from the moment we inherit from an Interface we are doomed to local interface indirections ? Thinking in C++, the only way to achieve hand-coded performance is to inherit from nothing, and the expressions assume that every member implements the value(int i) function. Then these functions are all easily inlined at compile time. I am not sure that the same thing is possible when we involve abstract classes even if they are declared static.
